### 🚀 **Classifying Images with MobileNetV2**

### 📦 Importing Pre-trained MobileNetV2

In [12]:
from keras.applications.mobilenet_v2 import MobileNetV2

model = MobileNetV2(weights="imagenet")
model.summary()

Model: "mobilenetv2_1.00_224"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_2       │ (None, 224, 224,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1 (Conv2D)      │ (None, 112, 112,  │        864 │ input_layer_2[0]… │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bn_Conv1            │ (None, 112, 112,  │        128 │ Conv1[0][0]       │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1_relu (ReLU)   │ (None, 112, 112,  │          0 │ bn_Conv1[0][0]    │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        288 │ Conv1_relu[0][0]  │
│ (DepthwiseConv2D)   │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        128 │ expanded_conv_de… │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │          0 │ expanded_conv_de… │
│ (ReLU)              │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │        512 │ expanded_conv_de… │
│ (Conv2D)            │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │         64 │ expanded_conv_pr… │
│ (BatchNormalizatio… │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand      │ (None, 112, 112,  │      1,536 │ expanded_conv_pr… │
│ (Conv2D)            │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_BN   │ (None, 112, 112,  │        384 │ block_1_expand[0… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_relu │ (None, 112, 112,  │          0 │ block_1_expand_B… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_pad         │ (None, 113, 113,  │          0 │ block_1_expand_r… │
│ (ZeroPadding2D)     │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise   │ (None, 56, 56,    │        864 │ block_1_pad[0][0] │
│ (DepthwiseConv2D)   │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │        384 │ block_1_depthwis… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │          0 │ block_1_depthwis… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_project     │ (None, 56, 56,    │      2,304 │ block_1_depthwis

 Total params: 3,538,984 (13.50 MB)

 Trainable params: 3,504,872 (13.37 MB)

 Non-trainable params: 34,112 (133.25 KB)

### 📷 Loading the image

In [13]:
import numpy as np
from imageio import imread

data = np.empty((1, 224, 224, 3)) # 1 batch, 224x224x3
data[0] = imread("images/cat.jpg")

/var/folders/qq/j2sb577j0w75h4b_p9m06xww0000gn/T/ipykernel_24521/1013612307.py:5: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  data[0] = imread("images/cat.jpg")


In [14]:
print(data[0]) # The image is not normalized
print(f'min: {data[0].min()}, max: {data[0].max()}')

[[[189. 173. 160.]
  [189. 173. 160.]
  [189. 173. 160.]
  ...
  [198. 178. 167.]
  [198. 178. 167.]
  [198. 178. 167.]]

 [[189. 173. 160.]
  [189. 173. 160.]
  [189. 173. 160.]
  ...
  [198. 178. 167.]
  [198. 178. 167.]
  [198. 178. 167.]]

 [[189. 173. 160.]
  [189. 173. 160.]
  [189. 173. 160.]
  ...
  [198. 178. 167.]
  [198. 178. 167.]
  [198. 178. 167.]]

 ...

 [[123. 125. 120.]
  [120. 122. 117.]
  [119. 121. 116.]
  ...
  [172. 167. 164.]
  [172. 167. 164.]
  [172. 167. 164.]]

 [[123. 123. 123.]
  [120. 120. 120.]
  [119. 119. 119.]
  ...
  [170. 159. 157.]
  [170. 159. 157.]
  [170. 159. 157.]]

 [[122. 122. 124.]
  [119. 119. 121.]
  [119. 119. 121.]
  ...
  [173. 158. 155.]
  [173. 158. 155.]
  [174. 159. 156.]]]
min: 0.0, max: 255.0


### 🔄 Normalizing the image

In [15]:
from keras.applications.mobilenet_v2 import preprocess_input

data = preprocess_input(data)

In [16]:
print(data[0]) # The image is normalized
print(f'min: {data[0].min()}, max: {data[0].max()}')

[[[ 0.48235294  0.35686275  0.25490196]
  [ 0.48235294  0.35686275  0.25490196]
  [ 0.48235294  0.35686275  0.25490196]
  ...
  [ 0.55294118  0.39607843  0.30980392]
  [ 0.55294118  0.39607843  0.30980392]
  [ 0.55294118  0.39607843  0.30980392]]

 [[ 0.48235294  0.35686275  0.25490196]
  [ 0.48235294  0.35686275  0.25490196]
  [ 0.48235294  0.35686275  0.25490196]
  ...
  [ 0.55294118  0.39607843  0.30980392]
  [ 0.55294118  0.39607843  0.30980392]
  [ 0.55294118  0.39607843  0.30980392]]

 [[ 0.48235294  0.35686275  0.25490196]
  [ 0.48235294  0.35686275  0.25490196]
  [ 0.48235294  0.35686275  0.25490196]
  ...
  [ 0.55294118  0.39607843  0.30980392]
  [ 0.55294118  0.39607843  0.30980392]
  [ 0.55294118  0.39607843  0.30980392]]

 ...

 [[-0.03529412 -0.01960784 -0.05882353]
  [-0.05882353 -0.04313725 -0.08235294]
  [-0.06666667 -0.05098039 -0.09019608]
  ...
  [ 0.34901961  0.30980392  0.28627451]
  [ 0.34901961  0.30980392  0.28627451]
  [ 0.34901961  0.30980392  0.28627451]]

 [

### ✅ Classifying the image

In [17]:
predictions = model.predict(data)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 345ms/step


In [18]:
predictions.shape # 1 batch, 1000 classes

(1, 1000)

In [19]:
np.argmax(predictions, axis=1) # Index of the maximum value

array([281])

In [20]:
predictions[0][281] # Probability of the 281th class

0.678309

### ⚙️ Decoding the predictions

In [21]:
from keras.applications.mobilenet_v2 import decode_predictions

for name, desc, score in decode_predictions(predictions, top=5)[0]:
  print('- {} ({:.2f}%)'.format(desc, 100 * score))

- tabby (67.83%)
- tiger_cat (16.14%)
- Egyptian_cat (5.40%)
- carton (0.36%)
- washbasin (0.33%)


### 🔚 Conclusion
According to the prediction results of the model, the provided ***cat.jpg*** image identifies a **tabby cat** with a percentage of **67.83%**.